service/history/historyEngine.go

```diff
@@ -585,44 +585,98 @@ func (e *historyEngineImpl) GetMutableState(
 func (e *historyEngineImpl) QueryWorkflow(
 	ctx ctx.Context,
 	request *h.QueryWorkflowRequest,
-) (*h.QueryWorkflowResponse, error) {
+) (retResp *h.QueryWorkflowResponse, retErr error) {
+	domainCache, err := e.shard.GetDomainCache().GetDomainByID(request.GetDomainUUID())
```

Comment:
i would suggest 
1. moving this logic into a dedicated file.
2. breaking the 3 logic (loading mutable state and do something) into 3 functions
3. it is ok to have a for loop for retry, but plz move the logic within the for loop into a function